### ISSUE TITLE
Reactor will not stop polling if not all active task queues has more tasks than the max backlog

### ISSUE THREAD

**Issue by denesb** [scylladb/seastar#652]
Normally, when the reactor has too many tasks (> than the configured _max_task_backlog) the reactor stops polling so it won't generate new ones and instead can focus on retiring the existing ones. However when there are multiple scheduling groups (task queues) this doesn't work. The problem is that when deciding whether it can poll it uses only the task count of the currently selected task queue. So when moving to a task queue with less then the limit tasks it will consider that it is okay to poll. Consider the following example:
```
task queue    backlog
A             100
B             2000
```
And a _max_task_backlog of 1000. Suppose the reactor is executing tasks from queue B when the task quota expires and g_need_preempt is set. Since the amount of tasks is above the limit it will reset g_need_preempt to false, so that polling is skipped and so that the reactor will continue executing tasks. Now suppose the scheduler now decides its time to run task queue A. After being done with the task-quota it will again look at the limit. Since the number of tasks in queue A is below the limit it decides it is okay to poll. This can possibly generate tasks for queue B. If the rate of tasks generated from polling is higher than the rate of retired tasks this will result in the B queue exploding, ultimately leading to OOM if this is maintained for a long enough time.

**Comment by avikivity** [comment]
Interesting. Seems like we can change the condition to consider the total number of queued tasks (and add a variable to hold it).

But the application should also be more careful not to enter this state.

**Comment by denesb** [comment]
> But the application should also be more careful not to enter this state.

Of course. This mechanism should be the last line of defence, when all else fails.

**Comment by gleb-cloudius** [comment]
If I read the code correctly the scenario above cannot happen. The code will not switch to the queue A until B has less than max_backlog tasks in it (both scheduling and polling are disabled), so what it may cause it big task quota violation for the queue B, but not much more polling.

**Comment by denesb** [comment]
Indeed it seems so. But even so we can have `num_scheduling_groups * (_max_backlog - 1)` tasks and happily continue polling. So there must be another explanation for the ~1M task queue I saw. Do we have a mechanism to penalize scheduling groups that violated their task quota?

**Comment by gleb-cloudius** [comment]
On Mon, Jul 22, 2019 at 03:27:11AM -0700, Botond Dénes wrote:
> Indeed it seems so. But even so we can have `num_scheduling_groups * (_max_backlog - 1)` tasks and happily continue polling.
> 
Yes, but since _max_backlog is more or less a random number
multiplying it by a fixed small number should not hurt much.

--
			Gleb.

**Comment by denesb** [comment]
>> On Mon, Jul 22, 2019 at 03:27:11AM -0700, Botond Dénes wrote: Indeed it seems so. But even so we can have `num_scheduling_groups * (_max_backlog - 1)` tasks and happily continue polling.

> Yes, but since _max_backlog is more or less a random number multiplying it by a fixed small number should not hurt much.

We now have user created scheduling groups too. Although AFAIK we also have a hard upper cap (16?) on the number of groups, so the end result should still not be too crazy.

**Comment by psarna** [comment]
I'll revive this thread from its eternal rest in the name of the overload project. Can somebody confirm/debunk that the theoretical scenario can lead to exceeding the max task limit per queue?

Let's assume that we have a single task responsible for scheduling 100 sub-tasks during its execution - for simplicity, let's say it's implemented as follows:
```cpp
future<> background_job();

future<> i_produce_lots_of_tasks() {
    for (int i = 0; i < 100; ++i) {
        (void)background_job();
    }
    return make_ready_future<>();
}
```

Now, let's also assume that we arrived at a state where our task queue is full of 1024 such tasks.

Isn't it pretty much guaranteed with our current implementation that we'll end up with 100k tasks in the queue, and now need to deal with the fact that these 100k tasks can also schedule more tasks on their way? Eventually the queue will dry up, but we may reach orders of magnitude more tasks than the initial limit.

While at it, it looks like we're in the mode of not switching to another queue until we dry this one up enough to go below the limit, but nothing prevents us from self-filling, possibly in an exponential fashion. And it's hard to do anything about it, because each task in the queue has a weight of 1, regardless if it schedules thousands of subtasks during its execution.

Does this hypothesis make sense, or did I miss something (which I personally find likely)?

Ref:
https://github.com/scylladb/seastar/blob/07758294ef78798980adc8f59c6bdc99405b8930/src/core/reactor.cc#L2220-L2247

**Comment by avikivity** [comment]
Your scenario will exhaust all memory, but there's nothing the reactor can do about it. Tasks that spawn other tasks should either employ some mechanism to limit concurrency, or promise that their branching factor is limited and reasonable.

**Comment by avikivity** [comment]
(two examples: a protocol server that supports pipelining has to limit concurrency, since the concurrency it generates is unbounded, but if processing a request has a concurrency of at least 3, then the request processor can consider its branching factor limited not add more controls)

**Comment by psarna** [comment]
> Your scenario will exhaust all memory, but there's nothing the reactor can do about it. Tasks that spawn other tasks should either employ some mechanism to limit concurrency, or promise that their branching factor is limited and reasonable.

Or, tasks could try to declare upfront how heavy their chains are, and the reactor could use these weights to take more than 1 unit from the semaphore if a task is heavy. But I agree that it doesn't sound like a feasible solution and I don't even imagine how such an API could look like.

**Comment by avikivity** [comment]
Tasks are too low-level to have such an interface. Remember tasks are generated from continuations and coroutines.

So such tasks are implicitly promising to be well behaved. Sources of concurrency need to declare themselves, and everyone else gets a free pass. Of course, we can try to provide better APIs for concurrency generators.

**Comment by eliransin** [comment]
I don't fully understand the reason for:
https://github.com/scylladb/seastar/blob/07758294ef78798980adc8f59c6bdc99405b8930/src/core/reactor.cc#L2237

The only solid reason I can think of for this particular condition is, to minimize the memory consumed by tasks that are waiting on the queue.
But then as @avikivity already mentioned, maybe we should consider the backlog as a whole (total of all queues). IIUC the reasoning is that we don't want to generate any more tasks if the memory is already filled by a lot of tasks.

About the unlimited concurrency that tasks can generate, if the average branching is greater than 1 and even 1 once we are over the backlog size, we will get stuck in this loop until we exhaust all of the memory. 
Maybe we should adjust some of the code to warn against that or help avoiding this:
1. We can easily collect stats per queue (scheduling group) about the average task branching, we just calculate the sub of the queue size before a task executed and after it has executed. (if it is larger than 1 we know we are in a divergence scenario). Once we do, we can issue warnings or just advertise the metric ourselves.
2. Pushing for a more "concurrency aware" API usage for example: to encourage using `max_concurrent_for_each`  instead of `parallel_for_each` by issuing a compilation warning maybe?

**Comment by avikivity** [comment]
Average branching greater than 1 is equivalent to infinite recursion. There's nothing the reactor can do to stop it.

This check is meant as a last-resort defense against unbounded concurrency. It limits concurrency by not admitting new I/O. It cannot do anything about already admitted concurrency, and if there's a bug that causes infinite tasks to be generated, it won't fix that bug.

**Comment by denesb** [comment]
So it seems the original problem I reported doesn't exist after all and the one @psarna reported later is not one we want to solve at this level, so I'm closing this issue. Feel free to re-open if you disagree.

**Comment by avikivity** [comment]
I think there is a weakness in the system. If sg A is appending tasks to sg B, and if B has few shares (or alternatively those tasks are expensive), then it's easy for A to keep B above the limit, ruining scheduling for everyone.

Perhaps with_scheduling_group(B) should block if B is overloaded. This will move the problem back to A.

I don't know how important this problem is.
